### PR TITLE
Tracking PR: Variable Hoisting 

### LINKED ISSUE #17: Variable Hoisting
Currently variables are only available after they are defined, as Passerine has no late-bound 'global scope' so to speak. However this makes things like mutual recursion a little annoying:

```passerine
-- TODO: remove this explicit hoist
odd = ()

even = n -> if (n == 0.0) {
    true
} else {
    odd (n - 1.0)
}

odd = n -> if (n == 0.0) {
    false
} else {
    even (n - 1.0)
}

print even 35.0
```
We have to declare odd before using it in even. Ideally this is something that can be resolved at compile time. I've started work on this in the `dev` branch by adding a step that resolves the scope of all locals / captured before compilation, but there are still a few kinks to work out. 

For those interested, here's how the system works:

Each time a lambda is introduced, a new scope is entered:

```passerine
-- outer scope
x = true

capture_function = () -> {
    y = 3.14
    -- inner scope
    print x
    print y
}
```

if a variable (i.e. `x`) is defined in the outer scope, it's accessible in the inner scope. This works by 'capturing' `x` when `capture_function` is defined (basically just making it a shared reference on the heap).

While compiling, we can imagine this as a chain of scopes. Just after the expression `print y`, the scope stack looks like this:

```rust
[ 
    Scope { locals: ["x"], nonlocals: [] }, 
    Scope { locals: ["y"], nonlocals: ["x"] },
]
```

Currently, whenever a variable is encountered, here's what currently happens:

1. If the variable is local to the scope, add it to `locals` in the current `Scope`
2. If the variable is not local to the scope, search backwards through all enclosing scopes. If it is found and add it to the `nonlocals` of all the scopes we've searched through
3. If the variable can not be found:
    a. If this is a declaration, i.e. `x = true` or `x -> {}` define the local in the current scope
    b. If this is an access, i.e `print x`, raise the error 'variable referenced before assignment'

This works well in most cases: it's able to hoist captured variables, declare new variables, etc. etc.

However, there are a few problems with this methodology. Consider this simple example:

```passerine
recurse = x -> recurse x
```

At the point in time where the expression is compiling `recurse ()`, the scope stack looks like:

```rust
[
    Scope { locals: [], nonlocals: [] }, // This is empty!
    Scope { locals: ["x"], nonlocals: [] }, // no reference to `recurse`
]
```

So this code will throw the error 'variable recurse referenced before assignment'.

How do we fix this? We do something a bit more complicated. Steps 1 - 3a are the same: let's focus on 3b:

> b. If this is an access, i.e `print x`, raise the error 'variable referenced before assignment'

`x`, for instance, can be defined after we parse the expression. consider the `recurse` example, or the first example with mutual recursion. What we need to do is somehow mark `x` as used, and when `x` is later defined, join the later assignment.

Let me introduce the solution: Hoisting chains! 😱 🎉 

So, here's the skinny, 3b becomes:

> b. If this is an access, i.e `print x`, add `x` to a table called `unresolved_captures`, and mark `x` as captured a la rule 2 in all enclosing scopes.

This second bit, 'mark `x` as captured' builds up what I like to refer to as a 'hoist chain'. We need one more change to make this work, 3a becomes:

> a. If this is a declaration, i.e. `x = true` or `x -> {}` check `unresolved_captures` for `x`.
>   i. If `x` is in unresolved captures, define it in the local scope, *remove* it from `unresolved_captures` and then *remove* it as captured in all enclosing scopes (a la rule 2).

An finally, we need an additional rule, rule 4, for dealing with cases where variables are actually undefined:

> 4. After compilation, if there are still variables in `unresolved_captures`, raise the error 'variable(s) referenced before assignment' (obviously point out which variables and where this occurred).

What does this do in practice? consider the following code:

```passerine
flubber = () -> {
    foo = () -> {
        print (pi + e)
    }
    pi = 3.14
    foo ()
}
e = 2.72

flubber ()
```

Note that we have two variables that are 'referenced before assignment': `e` and `pi`.

After we reach `pi` in the expression `print (pi + e)`, the scope stack looks like this:

```rust
[
    Scope { locals: [], nonlocals: ["pi"] }, 
    Scope { locals: [], nonlocals: ["pi"] },
    Scope { locals: [], nonlocals: ["pi"] },
]
```

Note how we built a hoisting chain of `pi`s. `pi` has also been added to `unresolved_captures`. We repeat this when we reach `e`.

When we reach `pi = 3.14` in the enclosing scope, we see that `pi` is in `unresolved_captures`. We then remove it from `unresolved_captures` and from being captured in all enclosing scopes:

```rust
[
    Scope { locals: [], nonlocals: ["e"] }, 
    Scope { locals: ["foo", "pi"], nonlocals: ["e"] },
    // this innermost scope has been popped at this point,
    // but it is correct as we hoisted `pi` and `e` in it
]
```

Likewise, when we reach `e = 2.72`:

```rust
[
    Scope { locals: ["flubber", "e"], nonlocals: [] }, 
    // scope has also been popped, but stil correct
]
```

These scopes are packaged into the `AST` and then used during bytecode emmision to resolve variables. 

It's a tad bit more complicated than this, because we're also renaming variables as we do this, but more on that later. Anyway, this feature should be merged in 0.9.X or 0.10.0 depending.

### PR DESCRIPTION
**Pull request by slightknack**
I added support for variable hoisting in the compiler; I just need to clean up the code, add better error messages, and write out some documentation/examples. This is technically a breaking change, but the 0.10.0 release isn't ready yet, So I think that releasing it as a 0.9.X release candidate is appropriate as there would be no major change to end users. This PR will fix #17.

### CODE REVIEW

**Comment by slightknack** [comment]
Looks like I'll need to rewrite some tests and such.

**Comment by hhhapz** [comment]
and uhhh clean that commit history

**Comment by slightknack** [comment]
> and uhhh clean that commit history

I'll squash and merge, don't worry.

File: src/compiler/hoist.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+/// Replaces all symbols with unique identifiers;
+/// symbols by the same name in different scopes will get different identifiers.
+/// Also resolves closure captures and closure hoisting.
+pub fn hoist(cst: Spanned<CST>) -> Result<(Spanned<SST>, Scope), Syntax> {
+    let mut hoister = Hoister::new();
+    let sst = hoister.walk(cst)?;
+    let scope = hoister.scopes.pop().unwrap();
+
+    if !hoister.unresolved_hoists.is_empty() {
+        // TODO: Actual errors
```
Comment (hhhapz) [inline src/compiler/hoist.rs:29]: Period.

File: src/compiler/hoist.rs [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+/// Keeps track of:
+/// 1. Local and nonlocal variables in each scope.
+/// 2. All variables declared.
+/// 3. Variables that have been used but not declared.
+pub struct Hoister {
+    /// The unique local symbols in the current scope.
+    scopes: Vec<Scope>,
+    /// Maps integers (index in vector) to string representation of symbol.
+    // TODO: make it it's own type
```
Comment (hhhapz) [inline src/compiler/hoist.rs:53]: Period.

File: src/compiler/hoist.rs [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        Hoister {
+            scopes:            vec![Scope::new()],
+            symbol_table:      vec![],
+            unresolved_hoists: HashMap::new(),
+        }
+    }
+
+    /// Enters a new scope, called when entering a new function.
+    fn   enter_scope(&mut self)               { self.scopes.push(Scope::new()); }
+    /// Enters an existing scope, called when resolving variables
```
Comment (hhhapz) [inline src/compiler/hoist.rs:72]: Period.

File: src/compiler/hoist.rs [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+    /// Returns the topmost scope immutably.
+    fn borrow_local_scope(&self) -> &Scope {
+        let last = self.scopes.len() - 1;
+        &self.scopes[last]
+    }
+
+    /// Walks a `CST` to produce an `SST`.
+    /// This is fairly standard - hoisting happens in
+    ///`self.assign`, `self.lambda`, and `self.symbol`.
```
Comment (hhhapz) [inline src/compiler/hoist.rs:95]: Space.

File: src/compiler/hoist.rs [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+            unresolved_hoists: HashMap::new(),
+        }
+    }
+
+    /// Enters a new scope, called when entering a new function.
+    fn   enter_scope(&mut self)               { self.scopes.push(Scope::new()); }
+    /// Enters an existing scope, called when resolving variables
+    fn reenter_scope(&mut self, scope: Scope) { self.scopes.push(scope)         }
+
+    /// Exits the current scope, returning it.
```
Comment (hhhapz) [inline src/compiler/hoist.rs:75]: "If the Hoister is in the outermost scope, it returns None."

File: src/compiler/hoist.rs [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+        if let Some(scope) = self.exit_scope() {
+            let resolved = self.try_resolve(name);
+            self.reenter_scope(scope);
+            if let Some(unique_symbol) = resolved {
+                self.local_scope().nonlocals.push(unique_symbol);
+                return Some(unique_symbol);
+            }
+        }
+
+        // variable is not defined in this or enclosing scopes
```
Comment (hhhapz) [inline src/compiler/hoist.rs:191]: Capitalization and period.

File: src/compiler/hoist.rs [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        return None;
+    }
+
+    // 1. If the variable is local to the scope, add it to `locals` in the current `Scope`
```
Comment (hhhapz) [inline src/compiler/hoist.rs:195]: Period.

File: src/compiler/hoist.rs [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+
+    /// Returns the unique usize of a local symbol.
+    /// If a variable is referenced before assignment,
+    /// this function will define it in all lexical scopes
+    /// once this variable is discovered, we remove the definitions
+    /// in all scopes below this one.
+    fn resolve_assign(&mut self, name: &str, redeclare: bool) -> UniqueSymbol {
+        // if we've seen the symbol before but don't know where it's defined
+        if let Some(unique_symbol) = self.unresolved_hoists.get(name) {
+            // this is a definition; we've resolved it!
```
Comment (hhhapz) [inline src/compiler/hoist.rs:211]: Capitalization.

File: src/compiler/hoist.rs [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+    //     b. If this is an access, i.e `print x`, add `x` to a table called `unresolved_captures`, and mark `x` as captured a la rule 2 in all enclosing scopes.
+    // 4. After compilation, if there are still variables in `unresolved_captures`, raise the error 'variable(s) referenced before assignment' (obviously point out which variables and where this occurred).
+
+    /// Returns the unique usize of a local symbol.
+    /// If a variable is referenced before assignment,
+    /// this function will define it in all lexical scopes
+    /// once this variable is discovered, we remove the definitions
+    /// in all scopes below this one.
+    fn resolve_assign(&mut self, name: &str, redeclare: bool) -> UniqueSymbol {
+        // if we've seen the symbol before but don't know where it's defined
```
Comment (hhhapz) [inline src/compiler/hoist.rs:209]: Capitalization and period.

File: src/compiler/hoist.rs [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+        if let Some(unique_symbol) = self.unresolved_hoists.get(name) {
+            // this is a definition; we've resolved it!
+            let unique_symbol = *unique_symbol;
+            self.uncapture_all(unique_symbol);
+            self.unresolved_hoists.remove(name);
+            self.local_scope().locals.push(unique_symbol);
+            return unique_symbol;
+        }
+
+        // if we haven't seen the symbol before,
```
Comment (hhhapz) [inline src/compiler/hoist.rs:219]: Capitalization and period.

File: src/compiler/hoist.rs [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+        // we search backwards through scopes and build a hoisting chain
+        if !redeclare {
+            if let Some(unique_symbol) = self.try_resolve(name) { return unique_symbol; }
+        }
+
+        // if we didn't find it by searching backwards, we declare it in the current scope
```
Comment (hhhapz) [inline src/compiler/hoist.rs:225]: Capitalization and period.

File: src/compiler/hoist.rs [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+        let unique_symbol = self.new_symbol(name);
+        self.local_scope().locals.push(unique_symbol);
+        return unique_symbol;
+    }
+
+    /// This function wraps try_resolve,
+    /// but checks that the symbol is unresolved first.
+    fn resolve_symbol(&mut self, name: &str) -> UniqueSymbol {
+        // if we've seen the symbol before but don't know where it's defined
```
Comment (hhhapz) [inline src/compiler/hoist.rs:234]: Capitalization and period.

File: src/compiler/hoist.rs [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+        for expression in tuple {
+            expressions.push(self.walk(expression)?)
+        }
+
+        Ok(SST::Tuple(expressions))
+    }
+
+    /// Walks an assignment.
+    /// Delegates to `walk_pattern` for capturing.
+    /// Assignments can capture existing variables
```
Comment (hhhapz) [inline src/compiler/hoist.rs:279]: Period.

File: src/compiler/hoist.rs [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+    /// Walks a function call.
+    pub fn call(&mut self, fun: Spanned<CST>, arg: Spanned<CST>) -> Result<SST, Syntax> {
+        return Ok(SST::call(
+            self.walk(fun)?,
+            self.walk(arg)?,
+        ));
+    }
+}
+
+// #[cfg(test)]
```
Comment (hhhapz) [inline src/compiler/hoist.rs:315]: Debug tests?

Comment (hhhapz) [inline src/compiler/hoist.rs:315]: Debug tests?

File: src/compiler/rule.rs [hunk 15]
```diff
@@ -9,36 +9,36 @@ use crate::common::{
...
+    ast::{AST, ASTPattern, ArgPattern},
     syntax::Syntax
 };
 
 // TODO: immutably capture external values used by macro
 // TODO: add context for macro application
 // NOTE: add spans?
 
-/// When a macro is expanded, `AST` slices captured by the macro Argument Pattern
+/// When a macro is expanded, `AST` slices captured by the macro Argument CSTPattern
```
Comment (hhhapz) [inline src/compiler/rule.rs:20]: Maybe uncapitalize 'Are'?

File: src/compiler/rule.rs [hunk 16]
```diff
@@ -9,36 +9,36 @@ use crate::common::{
...
-/// A `Binding` relates a name (within an Argument Pattern),
+/// A `Binding` relates a name (within an Argument CSTPattern),
 /// to an `AST` slice.
 type Bindings = HashMap<String, Spanned<AST>>;
 
-/// A rule has an Argument Pattern and an `AST`.
-/// When a form matches the `ArgPat`,
+/// A rule has an Argument CSTPattern and an `AST`.
+/// When a form matches the `ArgPattern`,
 /// A set of bindings are produced,
```
Comment (hhhapz) [inline src/compiler/rule.rs:28]: Capitalization and period.

File: src/compiler/rule.rs [hunk 17]
```diff
@@ -154,7 +154,7 @@ impl Rule {
     }
 
     /// Resolves a symbol.
```
Comment (hhhapz) [inline src/compiler/rule.rs:156]: Uncapitalize 'We'.

File: src/compiler/rule.rs [hunk 18]
```diff
@@ -280,13 +280,13 @@ impl Rule {
             },
 
             // replace the variables in (argument) patterns
```
Comment (hhhapz) [inline src/compiler/rule.rs:282]: Capitalization.

File: src/compiler/rule.rs [hunk 19]
```diff
@@ -280,13 +280,13 @@ impl Rule {
...
+                AST::CSTPattern(Rule::expand_pattern(spanned, bindings)?.item)
             },
-            AST::ArgPat(arg_pat) => {
+            AST::ArgPattern(arg_pat) => {
                 let spanned = Spanned::new(arg_pat, tree.span.clone());
-                AST::ArgPat(Rule::expand_arg_pat(spanned, bindings)?.item)
+                AST::ArgPattern(Rule::expand_arg_pat(spanned, bindings)?.item)
             },
 
             // replace the variables in the patterns and the expression
```
Comment (hhhapz) [inline src/compiler/rule.rs:292]: Capitalization.

File: src/compiler/sst.rs [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct UniqueSymbol(pub usize);
+
+/// A pattern that mirrors the structure of some Data.
+#[derive(Debug, Clone, PartialEq)]
+pub enum SSTPattern {
+    Symbol(UniqueSymbol),
+    Data(Data),
+    Label(String, Box<Spanned<SSTPattern>>), // todo usize for label
```
Comment (hhhapz) [inline src/compiler/sst.rs:14]: Capitalization and period.

File: src/compiler/sst.rs [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+    Tuple(Vec<Spanned<SSTPattern>>),
+    // Where {
```
Comment (hhhapz) [inline src/compiler/sst.rs:16]: Debug comments?

File: src/compiler/sst.rs [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+    },
+    Label(String, Box<Spanned<SST>>),
+    Tuple(Vec<Spanned<SST>>),
+    FFI {
+        name:       String,
+        expression: Box<Spanned<SST>>,
+    },
+}
+
+// pub struct ScopeContext {
```
Comment (hhhapz) [inline src/compiler/sst.rs:83]: Debug comments?

File: tests/snippets/before_assign.pn [hunk 23]
```diff
@@ -0,0 +1,7 @@
+-- action: run
+-- outcome: trace
+
+-- TODO: should be syntax
```
Comment (hhhapz) [inline tests/snippets/before_assign.pn:4]: Period.

**Comment by slightknack** [comment]
Thanks for the review! One quick note:

For documentation comments, I use full sentences with periods and capitalization.
For non documentation comments, I use lowercase with little to no punctuation.

**Review by hhhapz — CHANGES_REQUESTED** [review]
Overall, the logic sounds good and I can't think of any specific edge cases that may have been omitted in the tests. I see a lot of inconsistency in documentation, and I know that's difficult to enforce when programming, but perhaps it can be valuable to look into something more consistent, as well as max column lengths. Sometimes I see comments that break to new lines between sentences in ~ 50 cols, other times around 80. Maybe this can be something you can standardize on as I am unsure what Rust suggests.

Finally, just a few capitalization and sentences that aren't stopped with periods. Maybe that can be something you can use a quick regex/grep to find any others you are missing on. I also see a few snippets of commented code, however am not sure if they have any significance, so I just flagged them regardless.

File: src/compiler/sst.rs (hunk 21 above)
Comment (slightknack) [reply src/compiler/sst.rs:16]: This is a planned feature; just have it commented out for now.

**Comment by slightknack** [comment]
> ... as well as max column lengths. Sometimes I see comments that break to new lines between sentences in ~ 50 cols, other times around 80. Maybe this can be something you can standardize on as I am unsure what Rust suggests.

Rust has the `rustfmt` tool for enforcing consistent formatting, including comment lengths and so on. However, I'm not a huge fan of the defaults, and it doesn't support aligning `match` arms:

```rust
match x {
    Some(thing) => thing,
    None        => 0,
    // will smush the arrow
}
```
I mean, it's pretty configurable, and supports aligning, i.e. struct fields; I just don't think the benefits outweigh the costs *yet*.

File: src/compiler/sst.rs (hunk 22 above)
Comment (slightknack) [reply src/compiler/sst.rs:83]: Yeah, will remove

File: src/compiler/hoist.rs (hunk 14 above)
Comment (slightknack) [reply src/compiler/hoist.rs:316]: Yeah

**Review by hhhapz — APPROVED** [review]
